### DISCUSSION TITLE
JUnit 5

### DISCUSSION THREAD

**Discussion by mgenereu** [com-lihaoyi/mill#1411] [Development]
I brought this up in gitter.im but I'm not sure this isn't a better place?

Has there been any conversation on junit 5 support from mill? The interface at https://github.com/sbt/sbt-jupiter-interface has an `sbt.testing.Framework` implementation but the fingerprint implementation is nerfed (and poorly at that). The accompanying plugin then provides the classes via SBTs `definedTests` and uses junit 5's Discovery mechanism because "SBT is currently not capable of detecting package private test classes." I'd like to get this working but I could use some mentorship on "the mill way" to accomplish this integration. I'm tempted to fork sbt-jupiter-interface, fix the nerfed implementation, and forgo the private test classes but that sounds evil.

Little background: I'm running Apache Camel from Mill and Camel provides support classes for Junit 4 and 5. I'm not sure what 5 offers over 4 but appears to be accepted as the future of that project.

**Comment by lefou** [comment]
If the issue which you describe as "nerfed" is fixable by fixing the implementation of `sbt.testing.Framework`, than I don't see why that "sounds evil", quite the opposite, it would be the best way to fix it. Mill also uses `sbt.testing.Framework` implementations to locate tests, so a proper fix, either as pull request to sbt-jupiter-interface or as (temporary) fork to see if it works, looks like the best way to go. And as such, it is completely decoupled from mill API but still the "mill way". ;-)

TBH, I didn't understand your issue completely, and I have no idea why the sbt plugin bypasses the sbt-testing interface, other than for the reason that sbt-testing interface is not capable enough. Maybe you can shed some more light in case I completely failed to see the point.

**Comment by mgenereu** [comment]
The nerfed implementation is so the SBT plugin, instead of the SBT Framework Library that uses Fingerprints and only Fingerprints, can talk to the JUnit 5 Discovery library and have it discover the Test Classes including private ones.  This can't be done with Fingerprints.

A separate issue I have with the Fingerprint nerfing is that the `annotationName` method returns a class name that's not an descendant of `Annotation`.  Returning a class name like that seems tacky and [mill's SBT Fingerprint scanner](https://github.com/com-lihaoyi/mill/blob/main/scalalib/src/Lib.scala#L179) casts the loading of that class as an `Annotation`.  Maybe I misread things but IDEA's debugger really didn't like it either.

**Reply by lefou** [reply]
First, each Java annotation is a `java.lang.annotation.Annotation`.

So, you say, that JUnit 5 tests can't be discovered by a locator contract such as `sbt.testing.Framework` because the available `Fingerprint`s aren't suitable to what is needed for JUnit 5? What's missing? You should open a issue/PR in sbt/test-interface repo, too.

**Reply by mgenereu** [reply]
Here's the line that I have issue with: https://github.com/sbt/sbt-jupiter-interface/blob/dc737ac36eaacb585eec069b91d2c6520fb983d9/src/library/src/main/java/net/aichler/jupiter/api/JupiterTestFingerprint.java#L46

**Comment by mgenereu** [comment]
Maybe it's just a matter of extending Mill to allow me to provide an alternative implementation for `discoverTests` that can then call the JUnit Discovery mechanism like the SBT plugin does.  That way the nerfed implementation never gets called and we can still use the rest of the `sbt-jupiter-interface` Framework.

**Comment by ice1000** [comment]
> TBH, I didn't understand your issue completely

I think the OP's question can be simplified as 'how to use junit5 with mill'

**Reply by lefou** [reply]
The simple answer is: Provide a proper implementation of `sbt.testing.Framework`.

Looks like there already exists one, but it is in some ways limitted. So, in summary, either fix the existing implementation, or provide a better one.

Beside providing a framework implenentation, there is no other work needed to run those tests in mill.

As an alternative, if there is some standalone runner for Junit 5 tests, you can also just invoke this, but I haven't looked into Junit 5, so I can't tell you any details.

**Reply by mgenereu** [reply]
The JUnit 5 Test class discovery mechanism cannot be implemented using `sbt.testing.Framework` since the discovery isn't Fingerprint based.

**Reply by lefou** [reply]
It's probably a good idea to open an issue in https://github.com/sbt/test-interface. I could imagine, that it can be extended to also work for JUnit 5. If you already can provide details, how to discover test cases in JUnit 5, the better.

**Reply by mgenereu** [reply]
Got it. Thank you!
